### PR TITLE
vm: allow to emit stackitem.Convertible

### LINKED ISSUE #2987: notary: Example deposit code is currently unworkable
I try to make notary deposit using https://pkg.go.dev/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/nep17#TokenWriter.Transfer.

I used https://pkg.go.dev/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/notary#OnNEP17PaymentData as extra GAS transfer data which conveniently groups parameters. This is also used in https://pkg.go.dev/github.com/nspcc-dev/neo-go@v0.101.1/pkg/rpcclient/notary#pkg-examples.

Sad thing that `Transfer` method doesn't support `OnNEP17PaymentData` type. I realize that example skip error conditions but I'd expect code to be usable in good conditions.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #2987.

### CODE REVIEW

File: pkg/rpcclient/notary/contract.go [hunk 1]
```diff
@@ -234,3 +240,40 @@ func withdrawScript(from util.Uint160, to util.Uint160) []byte {
...
+		if err != nil {
+			return fmt.Errorf("failed to decode acctount bytes: %w", err)
+		}
+		d.Account = &acc
+	}
+	till, err := arr[1].TryInteger()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve till: %w", err)
+	}
+	d.Till = uint32(till.Int64())
```
Comment (roman-khimov) [inline pkg/rpcclient/notary/contract.go:277]: What if it's `!till.IsInt64()`? What if it's larger than `math.MaxUint32`?

File: pkg/vm/emit/emit.go [hunk 2]
```diff
@@ -160,6 +164,54 @@ func Array(w *io.BinWriter, es ...any) {
...
+		}
+		Array(w, arrAny...)
+	case stackitem.MapT:
+		// TODO: this one is very controversial, I vote for returning an error here.
+		arr := si.Value().([]stackitem.MapElement)
+		arrAny := make([]any, len(arr))
+		for i := range arr {
+			arrAny[i] = &arr[i] // *stackitem.MapElement is Convertible.
+		}
+		Array(w, arrAny...)
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:208]: Array is not a map.

File: pkg/vm/stackitem/item.go [hunk 3]
```diff
@@ -799,6 +799,25 @@ type MapElement struct {
 	Value Item
 }
 
+// ToStackItem implements convertible interface and is used for emitter compatibility.
```
Comment (roman-khimov) [inline pkg/vm/stackitem/item.go:802]: I don't think you really need it.

File: pkg/rpcclient/notary/contract.go (hunk 1 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/notary/contract.go]: I thought about it, but Notary can't properly handle these data, thus I've disided to skip these checkc. Although you're right and these methods should work with any data.

Fixed.

File: pkg/vm/emit/emit.go (hunk 2 above)
Comment (AnnaShaleva) [reply pkg/vm/emit/emit.go]: I've used `PACKSTRUCT` and `PACKMAP` so rarely that I completely forgot that they exist. These two opcodes solve the problem, fixed.

File: pkg/vm/stackitem/item.go (hunk 3 above)
Comment (AnnaShaleva) [reply pkg/vm/stackitem/item.go]: With `PACKMAP` we don't need it, that's true.

File: pkg/rpcclient/notary/contract.go [hunk 4]
```diff
@@ -234,3 +241,47 @@ func withdrawScript(from util.Uint160, to util.Uint160) []byte {
...
+	till, err := arr[1].TryInteger()
+	if err != nil {
+		return fmt.Errorf("failed to retrieve till: %w", err)
+	}
+	if !till.IsInt64() {
+		return errors.New("till is not an int64")
+	}
+	val := till.Int64()
+	if val > math.MaxUint32 {
+		return fmt.Errorf("till is larger than ma uint32 value: %d", val)
```
Comment (roman-khimov) [inline pkg/rpcclient/notary/contract.go:283]: ma?

File: pkg/rpcclient/notary/contract.go [hunk 5]
```diff
@@ -234,3 +241,47 @@ func withdrawScript(from util.Uint160, to util.Uint160) []byte {
...
+	if arr[0] != stackitem.Item(stackitem.Null{}) {
+		accBytes, err := arr[0].TryBytes()
+		if err != nil {
+			return fmt.Errorf("failed to retrieve account bytes: %w", err)
+		}
+		acc, err := util.Uint160DecodeBytesBE(accBytes)
+		if err != nil {
+			return fmt.Errorf("failed to decode acctount bytes: %w", err)
+		}
+		d.Account = &acc
```
Comment (roman-khimov) [inline pkg/rpcclient/notary/contract.go:272]: Please improve coverage for this.

File: pkg/rpcclient/notary/contract.go (hunk 4 above)
Comment (AnnaShaleva) [reply pkg/rpcclient/notary/contract.go]: Should be `max`, fixed.

**Comment by AnnaShaleva** [comment]
I'll add some more tests.

File: pkg/vm/emit/emit.go [hunk 6]
```diff
@@ -119,6 +119,8 @@ func Array(w *io.BinWriter, es ...any) {
 			Array(w, e...)
 		case int64:
 			Int(w, e)
+		case uint64:
+			Int(w, int64(e))
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:123]: Have you tried pushing `math.MaxUint64` this way?

Comment (AnnaShaleva) [reply pkg/vm/emit/emit.go:123]: Oh, my bad, I kept in mind that they have the same size and conversion will be successful, but the result will be far from the user's expectations.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**